noir_stdlib/src/hash/sha256.nr

```diff
@@ -169,99 +169,352 @@ unconstrained fn build_msg_block<let N: u32>(
     } else {
         BLOCK_SIZE
     };
-    for k in 0..block_input {
-        msg_block[k] = msg[msg_start + k];
+
+    // Figure out the number of items in the int array that we have to pack.
+    // e.g. if the input is [0,1,2,3, 4,5] then we need to pack it as 2 items: [0123, 4500]
+    let mut int_input = block_input / INT_SIZE;
+    if block_input % INT_SIZE != 0 {
+        int_input = int_input + 1;
+    };
+
+    for i in 0..int_input {
+        let mut msg_item: u32 = 0;
+        // Always construct the integer as 4 bytes, even if it means going beyond the input.
+        for j in 0..INT_SIZE {
+            let k = i * INT_SIZE + j;
+            let msg_byte = if k < block_input {
+                msg[msg_start + k]
+            } else {
+                0
+            };
+            msg_item = lshift8(msg_item, 1) + msg_byte as u32;
+        }
+        msg_block[i] = msg_item;
     }
+
+    // Returning the index as if it was a 64 byte array.
+    // We have to project it down to 16 items and bit shifting to get a byte back if we need it.
     (msg_block, block_input)
 }
 
 // Verify the block we are compressing was appropriately constructed by `build_msg_block`
 // and matches the input data. Returns the index of the first unset item.
+// If `message_size` is less than `msg_start` then this is called with the old non-empty block;
+// in that case we can skip verification, ie. no need to check that everything is zero.
 fn verify_msg_block<let N: u32>(
     msg: [u8; N],
     message_size: u32,
     msg_block: MSG_BLOCK,
     msg_start: u32,
 ) -> BLOCK_BYTE_PTR {
-    let mut msg_byte_ptr: u32 = 0; // Message byte pointer
+    let mut msg_byte_ptr = 0;
     let mut msg_end = msg_start + BLOCK_SIZE;
     if msg_end > N {
         msg_end = N;
     }
+    // We might have to go beyond the input to pad the fields.
+    if msg_end % INT_SIZE != 0 {
+        msg_end = msg_end + INT_SIZE - msg_end % INT_SIZE;
+    }
 
-    for k in msg_start..msg_end {
-        if k < message_size {
-            assert_eq(msg_block[msg_byte_ptr], msg[k]);
+    // Reconstructed packed item.
+    let mut msg_item: u32 = 0;
+
+    // Inclusive at the end so that we can compare the last item.
+    let mut i: u32 = 0;
+    for k in msg_start..=msg_end {
+        if k % INT_SIZE == 0 {
+            // If we consumed some input we can compare against the block.
+            if msg_start < message_size & k > msg_start {
```

Comment:
Noir wants parens for if conditional statements 🤔 surprised it isn't yelling at you here, I'm wondering if the correct thing is being checked.
```suggestion
            if (msg_start < message_size) & (k > msg_start) {
```